vm/interpreter/src/lib.rs

```diff
@@ -1,130 +1,314 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+use actor::{ACCOUNT_ACTOR_CODE_ID, REWARD_ACTOR_ADDR};
 use address::Address;
 use blocks::Tipset;
 use clock::ChainEpoch;
 use default_runtime::{internal_send, DefaultRuntime};
+use forest_encoding::Cbor;
 use ipld_blockstore::BlockStore;
+use log::warn;
 use message::{Message, MessageReceipt, SignedMessage, UnsignedMessage};
 use num_bigint::BigUint;
-use num_traits::cast::ToPrimitive;
-use runtime::DefaultSyscalls;
-use vm::ActorError;
-use vm::{ExitCode, Serialized, StateTree};
-
+use num_traits::Zero;
+use runtime::Syscalls;
+use vm::{price_list_by_epoch, ActorError, ExitCode, Serialized, StateTree};
 /// Interpreter which handles execution of state transitioning messages and returns receipts
 /// from the vm execution.
-pub struct VM<'a, ST: StateTree, DB: BlockStore> {
+pub struct VM<'a, ST, DB, SYS> {
     state: ST,
     store: &'a DB,
     epoch: ChainEpoch,
+    syscalls: SYS,
     // TODO: missing fields
 }
 
-impl<'a, ST: StateTree, DB: BlockStore> VM<'a, ST, DB> {
-    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch) -> Self {
+impl<'a, ST, DB, SYS> VM<'a, ST, DB, SYS>
+where
+    ST: StateTree,
+    DB: BlockStore,
+    SYS: Syscalls + Copy,
+{
+    pub fn new(state: ST, store: &'a DB, epoch: ChainEpoch, syscalls: SYS) -> Self {
         VM {
             state,
             store,
             epoch,
+            syscalls,
         }
     }
 
+    /// Returns ChainEpoch
+    fn epoch(&self) -> ChainEpoch {
+        self.epoch
+    }
+
     /// Apply all messages from a tipset
     /// Returns the receipts from the transactions.
     pub fn apply_tip_set_messages(
         &mut self,
         _tipset: &Tipset,
         msgs: &TipSetMessages,
-    ) -> Result<Vec<MessageReceipt>, String> {
+    ) -> Result<Vec<ApplyRet>, String> {
         let mut receipts = Vec::new();
 
         for block in &msgs.blocks {
             // TODO: verifiy ordering of message execution
 
             for msg in &block.bls_messages {
-                receipts.push(self.apply_message(msg, &block.miner)?);
+                receipts.push(self.apply_message(msg)?);
             }
 
             for msg in &block.secp_messages {
-                receipts.push(self.apply_message(&msg.message(), &block.miner)?);
+                receipts.push(self.apply_message(msg.message())?);
             }
         }
 
         Ok(receipts)
     }
 
+    fn _apply_implicit_message(&mut self, msg: &UnsignedMessage) -> Result<ApplyRet, String> {
```

Comment:
I'm confused as to why this is a Result. You already put the error into the ApplyRet struct and never return an Err via Result